core/cont/inc/ROOT/RConcurrentHashColl.hxx

```diff
@@ -0,0 +1,51 @@
+// Author: Danilo Piparo May 2018
+
+/*************************************************************************
+ * Copyright (C) 1995-2018, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RConcurrentHashColl
+#define ROOT_RConcurrentHashColl
+
+#include <ROOT/TRWSpinLock.hxx>
+
+#include <set>
+
+namespace ROOT {
+namespace Internal {
+
+/// This class is a TS set of unsigned set
+class RConcurrentUIntSet {
+private:
+   mutable std::set<unsigned int> fHashes;
+   mutable ROOT::TRWSpinLock fRWLock;
+public:
+   RConcurrentUIntSet() = default;
+   /// If the hash is there, return false. Otherwise, insert the hash and return true;
+   bool Insert(unsigned int hash) const
+   {
+      {
+#ifdef R__USE_IMT
```

Comment:
Why the conditional here?  Wouldn't it better to have this container always thread safe and use it only when IMT is on?

Reply:
sure.